torch/csrc/distributed/autograd/init.cpp

```diff
@@ -111,7 +111,12 @@ PyObject* dist_autograd_init(PyObject* /* unused */) {
         for (const auto& root : roots) {
           variables.emplace_back(root);
         }
-        DistEngine::getInstance().execute(variables);
+        try {
+          DistEngine::getInstance().execute(variables);
+        } catch (python_error & e) {
+          // FIXME: crashes if exception type is not RuntimeError
+          throw std::runtime_error(e.what());
```

Comment:
This was the cause of the distributed autograd failures. Without the exception translator, `python_error` exceptions were translated by `pybind11`'s default translator as a generic `std::exception`. I would consider that a bug since the original exception type is lost in the process. However, raising original exception seems to cause the worker to crash.

Reply:
What sort of failures did you see for distributed autograd? Also, regarding the `FIXME` statement, why does the worker crash if the exception type is not RuntimeError?

If possible it would be nice to add a unit test which repros the original problem and passes after this fix.

Reply:
>What sort of failures did you see for distributed autograd? 

The failure was in `test/dist_autograd_test.py`, specifically `test_backward_autograd_engine_error`. CI would hang and print some tracebacks for exceptions correctly propagated into the test and sometimes end with an error before hanging:
```
Exception: Simulate error on backward pass
terminate called without an active exception
```

>why does the worker crash if the exception type is not RuntimeError?

I'm not sure but I have tried with `Exception` and `ValueError` but only a `RuntimeError` seems to work. Presumably there is some assumption higher up the call stack being broken.

Reply:
@pritamdamania87, anything else to add?

Reply:
I'm fine with the change here to unblock this PR, but it would be nice to figure out why the worker crashed and if its possible to just throw the original exception.

Reply:
I've opened a separate issue for this: #32636